pkg/validation/test.go

```diff
@@ -522,23 +554,51 @@ func validateLiteralTestStep(context context, stage testStage, step api.LiteralT
 	}
 	if len(step.Commands) == 0 {
 		ret = append(ret, fmt.Errorf("%s: `commands` is required", context.fieldRoot))
+	} else {
+		ret = append(ret, validateCommands(step)...)
 	}
+
+	if step.BestEffort != nil && *step.BestEffort && step.Timeout == nil {
+		ret = append(ret, fmt.Errorf("test %s contains best_effort without timeout", step.As))
+	}
+
 	ret = append(ret, validateResourceRequirements(context.fieldRoot+".resources", step.Resources)...)
 	ret = append(ret, validateCredentials(context.fieldRoot, step.Credentials)...)
 	if err := validateParameters(&context, step.Environment); err != nil {
 		ret = append(ret, err)
 	}
 	ret = append(ret, validateDependencies(context.fieldRoot, step.Dependencies)...)
 	ret = append(ret, validateLeases(context.forField(".leases"), step.Leases)...)
-	switch stage {
-	case testStagePre, testStageTest:
+	switch validationArgs.testStage {
+	case TestStagePre, TestStageTest:
 		if step.OptionalOnSuccess != nil {
 			ret = append(ret, fmt.Errorf("%s: `optional_on_success` is only allowed for Post steps", context.fieldRoot))
 		}
 	}
 	return
 }
 
+func validateCommands(test api.LiteralTestStep) []error {
+	var validationErrors []error
+
+	var commands = test.Commands
+	scanner := bufio.NewScanner(strings.NewReader(commands))
+
+	hasTrapCommand := false
+	for scanner.Scan() {
```

Comment:
```suggestion
	if trapPattern.MatchString(test.Commands) {...
```

Reply:
Good call. I think I was initially going to try to do something very different from what I wound up doing here, but in its current state it definitely is not very useful to loop over each line.